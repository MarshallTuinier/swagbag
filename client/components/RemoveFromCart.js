import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Mutation } from "react-apollo";

import { CURRENT_USER_QUERY, REMOVE_FROM_CART_MUTATION } from "../GraphQL";

class RemoveFromCart extends Component {
  // This function will get called as soon as we get a response back from the server after a mutation
  update = (cache, payload) => {
    // Read the cache
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    // Remove the CartItem from the cart
    const cartItemId = payload.data.removeFromCart.id;
    data.currentUser.cart = data.currentUser.cart.filter(
      cartItem => cartItem.id !== cartItemId
    );
    // Rewrite to cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  };

  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        // Opitimistic UI
        optimisticResponse={{
          __typname: "Mutation",
          removeFromCart: {
            __typname: "CartItem",
            id: this.props.id
          }
        }}
      >
        {(removeFromCart, { loading, error }) => {
          return (
            <BigButton
              title="Delete Item"
              onClick={() => {
                removeFromCart().catch(err => alert(error.message));
              }}
              disabled={loading}
            >
              &times;
            </BigButton>
          );
        }}
      </Mutation>
    );
  }
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired
};

export default RemoveFromCart;

// -------------------- Styles -------------------- //

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${({ theme }) => theme.red};
    cursor: pointer;
  }
`;
