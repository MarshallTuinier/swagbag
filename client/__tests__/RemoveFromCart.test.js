import { mount } from "enzyme";
import { ApolloConsumer } from "react-apollo";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import toJSON from "enzyme-to-json";
import RemoveFromCart from "../components/RemoveFromCart";
import { REMOVE_FROM_CART_MUTATION, CURRENT_USER_QUERY } from "../GraphQL";
import { fakeUser, fakeCartItem } from "../lib/testUtils";

global.alert = console.log;

const user = fakeUser();
const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: { currentUser: { ...user, cart: [fakeCartItem({ id: "abc123" })] } }
    }
  },
  {
    request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: "abc123" } },
    result: {
      data: {
        removeFromCart: {
          __typename: "CartItem",
          id: "abc123"
        }
      }
    }
  }
];

describe("<RemoveFromCart />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider>
        <RemoveFromCart id="abc123" />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find("button"))).toMatchSnapshot();
  });

  it("removes the item from cart", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <RemoveFromCart id="abc123" />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    const {
      data: { currentUser }
    } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(currentUser.cart).toHaveLength(1);
    expect(currentUser.cart[0].item.price).toBe(5000);
    wrapper.find("button").simulate("click");
    await wait();
    const {
      data: { currentUser: updatedUser }
    } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(updatedUser.cart).toHaveLength(0);
  });
});
