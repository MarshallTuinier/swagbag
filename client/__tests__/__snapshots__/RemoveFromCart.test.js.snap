// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart__BigButton-sc-3q5ke3-0 hKuDcb"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Ã—
</button>
`;
